### PR TITLE
[FLINK-4616] Added functionality through which watermarks for each partition are saved and loaded via checkpointing mechanism

### PR DESCRIPTION
**Pull request by MayerRoman**
[FLINK-4616] Kafka consumer doesn't store last emmited watermarks per partition in state.

### CODE REVIEW

**Comment by MayerRoman** [comment]
I think that the changes that I propose eliminates the possibility of starting with checkpoints created before my code changes.

Because now it saves ListState<Tuple2<KafkaTopicPartition, Tuple2<Long, Long>>> (partition + offset + watermark).
And before it saved ListState<Tuple2<KafkaTopicPartition, Long>> (partition + offset).

(I mean checkpoints version later then 1.1.
Recently Added backward compatibility with 1.1 snapshots is taken into account in my commit with it, I think everything is ok)


Please advise me how to repair backward compatibility.

I have some idea of how to implement it:

1)	somehow verify returned from stateStore.getSerializableListState(..) object
in initializeState method
https://github.com/apache/flink/pull/3031/files?diff=unified#diff-06bf4a7f73d98ef91309154654563475R321

is it
ListState<Tuple2<KafkaTopicPartition, Long>>
or
ListState<Tuple2<KafkaTopicPartition, Tuple<Long, Long>>>

2)	Use for saving watermark separate state-object.

Or it is necessary implement different way.

I would be grateful for help.

**Comment by tzulitai** [comment]
Thank you for the contribution @MayerRoman. Just want to let you know that I've noticed this PR, and I think the issue is definitely something we want to fix. I'll allocate some time this week to review the PR.

**Comment by MayerRoman** [comment]
Thank you Tzu-Li Tai.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 1]
```diff
@@ -175,34 +176,115 @@ protected AbstractFetcher(
...
-		HashMap<KafkaTopicPartition, Long> state = new HashMap<>(allPartitions.length);
-		for (KafkaTopicPartitionState<?> partition : subscribedPartitions()) {
-			state.put(partition.getKafkaTopicPartition(), partition.getOffset());
+		HashMap<KafkaTopicPartition, Tuple2<Long, Long>> state = new HashMap<>(allPartitions.length);
+
+		switch (timestampWatermarkMode) {
+
+			case NO_TIMESTAMPS_WATERMARKS: {
+
+				for (KafkaTopicPartitionState<KPH> partition : allPartitions) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Excessive empty line above this line.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 2]
```diff
@@ -175,34 +176,115 @@ protected AbstractFetcher(
...
+				for (KafkaTopicPartitionStateWithPunctuatedWatermarks<T, KPH> partition : partitions) {
+					state.put(partition.getKafkaTopicPartition(), Tuple2.of(partition.getOffset(), partition.getCurrentPartitionWatermark()));
+				}
+
+				return state;
+			}
+
+			default:
+				// cannot happen, add this as a guard for the future
+				throw new RuntimeException();
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Would be good to have a reason message here.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 3]
```diff
@@ -175,34 +176,115 @@ protected AbstractFetcher(
...
+					if (watermarkTimestamp != null) {
+						partition.setCurrentWatermarkTimestamp(watermarkTimestamp);
+					}
+				}
+				break;
+			}
+
+			default:
+				// cannot happen, add this as a guard for the future
+				throw new RuntimeException();
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Would be good to have a reason message here.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionStateWithPeriodicWatermarks.java [hunk 4]
```diff
@@ -61,6 +61,10 @@ public long getCurrentWatermarkTimestamp() {
 		return partitionWatermark;
 	}
 
+	void setCurrentWatermarkTimestamp(long watermarkTimestamp) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionStateWithPeriodicWatermarks.java]: The other methods seem to be `public` (although they can actually be package-private). Should we stay consistent with that here?

**Comment by tzulitai** [comment]
A re-clarification about backwards compatibility for state type change:
Currently, one limitation for compatible applications across savepoint restore is that you can't change the type of state otherwise state restore will fail, therefore not compatible. The only work around, is to have another field as the new state with the new type, and somehow try to "encode" / "decode" the watermark state into / from the original `Tuple2<KafkaTopicPartition, Long>`. I don't think this is easily possible ...

At the same time, there was recent discussion about letting the window operators also checkpoint watermarks. So perhaps we might not need to let the Kafka sources checkpoint watermarks in the end, if the window operators already take care of restoring the previous event time.
What I'm curious about right now is whether or not in the future, redistributions of Kafka partition states across source subtasks will work well with the checkpointed watermarks in the downstream window operators. I don't think there should be a problem.

@aljoscha can you perhaps help clarify this?

**Comment by rmetzger** [comment]
I think this PR should also include a test for the added feature.

**Comment by MayerRoman** [comment]
I hope that I ended up with another issue, and I come back to this.

First, I want to ask a question that perhaps remove all the others.

Tzu-Li Tai, did I understand correctly that if discussion about letting the window operators checkpoint watermarks lead to the decision to implement this functionality in the window operators, the need to preserve the state of watermarks in Kafka consumer will disappear?

**Comment by tzulitai** [comment]
Hi @MayerRoman! Thank you for coming back to this issue.

I had a quick chat offline with @aljoscha about whether or not it'll be reasonable to add this. Either your approach in this PR or letting window operators checkpoint watermarks will both solve the issue of late elements after restore. We thought that we should let the user function code be free of responsibility of checkpointing watermarks, and let user code simply leave that to the streaming internals (checkpointed by window operators, and perhaps also by source operators).

So, essentially, the Kafka consumer should not need to checkpoint watermarks, and we can close this PR and the JIRA ticket. Very sorry for the late discussion on this, and having you worked on it already.

Let we know what you think and whether or not you agree :-)

**Comment by MayerRoman** [comment]
Hello, Tzu-Li Tai!
I think you make a good decision and agree to the fact that the PR and JIRA ticket can be closed.

Do not worry about the done work, I got good experience in the process.

**Comment by tzulitai** [comment]
Thanks. Can you please close this PR then :-D ? I'll close the JIRA.

**Comment by MayerRoman** [comment]
Ok

**Pull request closed without merging**